servers/src/common/types.rs

```diff
@@ -421,3 +424,94 @@ impl chain::TxHashsetWriteStatus for SyncState {
 		self.update(SyncStatus::TxHashsetDone);
 	}
 }
+
+/// A node is either "stem" of "fluff" for the duration of a single epoch.
+/// A node also maintains an outbound relay peer for the epoch.
+#[derive(Debug)]
+pub struct DandelionEpoch {
+	config: DandelionConfig,
+	// When did this epoch start?
+	start_time: Option<i64>,
+	// Are we in "stem" mode or "fluff" mode for this epoch?
+	is_stem: bool,
+	// Our current Dandelion relay peer (effective for this epoch).
+	relay_peer: Option<Arc<p2p::Peer>>,
+}
+
+impl DandelionEpoch {
+	/// Create a new Dandelion epoch, defaulting to "stem" and no outbound relay peer.
+	pub fn new(config: DandelionConfig) -> DandelionEpoch {
+		DandelionEpoch {
+			config,
+			start_time: None,
+			is_stem: true,
+			relay_peer: None,
+		}
+	}
+
+	/// Is the current Dandelion epoch expired?
+	/// It is expired if start_time is older than the configured epoch_secs.
+	pub fn is_expired(&self) -> bool {
+		let expired = if let Some(start_time) = self.start_time {
+			let epoch_secs = self.config.epoch_secs.expect("epoch_secs config missing") as i64;
+			Utc::now().timestamp().saturating_sub(start_time) > epoch_secs
+		} else {
+			true
+		};
+		expired
+	}
+
+	/// Transition to next Dandelion epoch.
+	/// Select stem/fluff based on configured stem_probability.
+	/// Choose a new outbound stem relay peer.
+	pub fn next_epoch(&mut self, peers: &Arc<p2p::Peers>) {
+		self.start_time = Some(Utc::now().timestamp());
+		self.relay_peer = peers.outgoing_connected_peers().first().cloned();
+
+		// If stem_probability == 90 then we stem 90% of the time.
+		let mut rng = rand::thread_rng();
+		let stem_probability = self
+			.config
+			.stem_probability
+			.expect("stem_probability config missing");
+		self.is_stem = rng.gen_range(0, 101) < stem_probability;
```

Comment:
Off by one? If I set a probability of 100, the condition should always be true, and in this case won't if the `rng` comes up with 100.

Reply:
🤦‍♂️ Yes. Should be `<=` here.

Reply:
```
rng.gen_range(0, 100) < stem_probability

# if stem probability is 0% then false at 0 and false at 99
# if stem probability is 100% then true at 0 and true at 99
```

Reply:
Not sure why I was having so much trouble getting my head around that...